quickwit-search/src/root.rs

```diff
@@ -0,0 +1,376 @@
+                        index1 += 1;
+                        merged_hits.push(hit2.clone());
+                        index2 += 1;
+                    }
+                }
+            }
+            (Ordering::Less, _) => {
+                let remaining = hits1
+                    .get(index1..hits1.len())
+                    .unwrap_or(&Vec::new())
+                    .to_vec();
+                merged_hits.extend(remaining);
+                break;
+            }
+            (_, Ordering::Less) => {
+                let remaining = hits2
+                    .get(index2..hits2.len())
+                    .unwrap_or(&Vec::new())
+                    .to_vec();
+                merged_hits.extend(remaining);
+                break;
+            }
+            _ => break,
+        };
+    }
+
+    Ok(merged_hits)
+}
+
+/// Perform a distributed search.
+/// It sends a search request over gRPC to multiple leaf nodes and merges the search results.
+pub async fn root_search(
+    search_request: &SearchRequest,
+    metastore: &dyn Metastore,
+    client_pool: &Arc<SearchClientPool>,
+) -> anyhow::Result<SearchResult> {
+    let start_instant = tokio::time::Instant::now();
+
+    // Create a job for leaf node search and assign the splits that the node is responsible for based on the job.
+    let split_metadata_list = list_relevant_splits(search_request, metastore).await?;
+
+    // Create a hash map of SplitMetadata with split id as a key.
+    let split_metadata_map: HashMap<String, SplitMetadata> = split_metadata_list
+        .into_iter()
+        .map(|split_metadata| (split_metadata.split_id.clone(), split_metadata))
+        .collect();
+
+    // Create a job for fetching docs and assign the splits that the node is responsible for based on the job.
+    let leaf_search_jobs: Vec<Job> = split_metadata_map
+        .keys()
+        .map(|split_id| {
+            // TODO: Change to a better way that does not use unwrap().
+            let split_metadata = split_metadata_map.get(split_id).unwrap();
+            Job {
+                split: split_id.clone(),
+                cost: compute_split_cost(split_metadata),
+            }
+        })
+        .collect();
+
+    let assigned_leaf_search_jobs = client_pool.assign_jobs(leaf_search_jobs).await?;
+
+    let clients = client_pool.clients.read().await;
+
+    // Create search request with start offset is 0 that used by leaf node search.
+    let mut search_request_with_offset_0 = search_request.clone();
+    search_request_with_offset_0.start_offset = 0;
+    search_request_with_offset_0.max_hits += search_request.start_offset;
+
+    // Perform the query phese.
+    let mut leaf_search_handles: Vec<JoinHandle<anyhow::Result<LeafSearchResult>>> = Vec::new();
+    for (addr, jobs) in assigned_leaf_search_jobs.iter() {
+        let mut search_client = if let Some(client) = clients.get(&addr) {
+            client.clone()
+        } else {
+            return Err(anyhow::anyhow!(
```

Comment:
nitpick (no need to change):
anyhow has a macro that is nice for that kind return.

```return Err(anyhow::anyhow!(..., addr))```

can be written
```bail!(..., addr)```